packages/job-worker/src/blueprints/context/services/PartAndPieceInstanceActionService.ts

```diff
@@ -0,0 +1,537 @@
+
+	public queuedPartInstanceId: PartInstanceId | undefined = undefined
+
+	constructor(
+		context: JobContext,
+		playoutModel: PlayoutModel,
+		showStyle: ReadonlyDeep<ProcessedShowStyleCompound>,
+		private readonly _rundown: PlayoutRundownModel
+	) {
+		this._context = context
+		this._playoutModel = playoutModel
+		this.showStyleCompound = showStyle
+	}
+
+	private _getPartInstance(part: 'current' | 'next'): PlayoutPartInstanceModel | null {
+		switch (part) {
+			case 'current':
+				return this._playoutModel.currentPartInstance
+			case 'next':
+				return this._playoutModel.nextPartInstance
+			default:
+				assertNever(part)
+				logger.warn(`Blueprint action requested unknown PartInstance "${part}"`)
+				throw new Error(`Unknown part "${part}"`)
+		}
+	}
+
+	async getPartInstance(part: 'current' | 'next'): Promise<IBlueprintPartInstance | undefined> {
+		const partInstance = this._getPartInstance(part)
+
+		return partInstance ? convertPartInstanceToBlueprints(partInstance.partInstance) : undefined
+	}
+	async getPieceInstances(part: 'current' | 'next'): Promise<IBlueprintPieceInstance[]> {
+		const partInstance = this._getPartInstance(part)
+		return partInstance?.pieceInstances?.map((p) => convertPieceInstanceToBlueprints(p.pieceInstance)) ?? []
+	}
+	async getResolvedPieceInstances(part: 'current' | 'next'): Promise<IBlueprintResolvedPieceInstance[]> {
+		const partInstance = this._getPartInstance(part)
+		if (!partInstance) {
+			return []
+		}
+
+		const resolvedInstances = getResolvedPiecesForCurrentPartInstance(
+			this._context,
+			this.showStyleCompound.sourceLayers,
+			partInstance
+		)
+		return resolvedInstances.map(convertResolvedPieceInstanceToBlueprints)
+	}
+
+	async findLastPieceOnLayer(
+		sourceLayerId0: string | string[],
+		options?: {
+			excludeCurrentPart?: boolean
+			originalOnly?: boolean
+			piecePrivateDataFilter?: any
+		}
+	): Promise<IBlueprintPieceInstance | undefined> {
+		const query: MongoQuery<PieceInstance> = {}
+		if (options?.piecePrivateDataFilter) {
+			for (const [key, value] of Object.entries<unknown>(options.piecePrivateDataFilter)) {
+				// TODO do we need better validation here?
+				// It should be pretty safe as we are working with the cache version (for now)
+				query[`piece.privateData.${key}`] = value
+			}
+		}
+
+		if (options?.excludeCurrentPart && this._playoutModel.playlist.currentPartInfo) {
+			query['partInstanceId'] = { $ne: this._playoutModel.playlist.currentPartInfo.partInstanceId }
+		}
+
+		const sourceLayerId = Array.isArray(sourceLayerId0) ? sourceLayerId0 : [sourceLayerId0]
+
+		const lastPieceInstance = await innerFindLastPieceOnLayer(
+			this._context,
+			this._playoutModel,
+			sourceLayerId,
+			options?.originalOnly ?? false,
+			query
+		)
+
+		return lastPieceInstance && convertPieceInstanceToBlueprints(lastPieceInstance)
+	}
+
+	async findLastScriptedPieceOnLayer(
+		sourceLayerId0: string | string[],
+		options?: {
```

Comment:
I would prefer for any method arguments in this class to be `| undefined` instead of optional, as it reduces the risk of forgetting to forward a newly added parameter which originates from the `IPartAndPieceActionContext` interface.

It would be bad and very confusing for a new parameter to be added in `IPartAndPieceActionContext`, implement it in `OnTakeContext` but forget to do the same in `AdlibExecutionContext`

Reply:
I updated four methods in this file